File: day18/main.go
package main

import (
	"bytes"
	_ "embed"
	"fmt"
	"math"
	"strconv"
	"sync"
	"sync/atomic"

	"github.com/solokirrik/aoc2023/ops"
)

//go:embed input
var input []byte

//go:embed input-example
var inputExample []byte

func main() {
	// 200ms
	ops.TimeIt(func() {
		inp := bytes.Split(input, []byte("\n"))
		entries := make([]Entry, 0, len(inp))
		for _, l := range inp {
			parts := bytes.Split(l, []byte(" "))
			dist := int(parts[1][0] - '0')
			if len(parts[1]) > 1 {
				dist = int(parts[1][0]-'0')*10 + int(parts[1][1]-'0')
			}
			e := Entry{
				Dir:  string(parts[0]),
				Dist: dist,
			}

			entries = append(entries, e)
		}

		out1 := Task1(entries)
		fmt.Println(76387 == out1, out1)
	})

	// heat death of the universe
	ops.TimeIt(func() {
		inp := bytes.Split(inputExample, []byte("\n"))
		entries := make([]Entry, 0, len(inp))
		for _, l := range inp {
			part2Info := bytes.Trim(bytes.Split(l, []byte(" "))[2], "()")
			e := Entry{
				Dist: decodeHexIntoDec(string(part2Info[1 : len(part2Info)-1])),
				Dir:  numToDir(int(part2Info[len(part2Info)-1]) - '0'),
			}
			entries = append(entries, e)
		}

		out2 := Task1(entries)
		fmt.Println(952408144115 == out2, out2)
	})
}

type Entry struct {
	Dir  string
	Dist int
}

func Task1(entries []Entry) int {
	perimetr, perimIndex, topLeft, downRight := buildPerimeter(entries)
	fmt.Println(topLeft, downRight)

	cubics := atomic.Uint64{}
	wg := sync.WaitGroup{}

	for y := topLeft.y; y < downRight.y+1; y++ {
		for x := topLeft.x; x < downRight.x+1; x++ {
			wg.Add(1)
			go func(x, y float64, wg *sync.WaitGroup) {
				defer wg.Done()
				if _, ok := perimIndex[Point2D{x, y}]; ok {
					cubics.Add(1)
					return
				}
				if PointInPolygonFast(Point2D{x, y}, perimetr) {
					cubics.Add(1)
				}
			}(x, y, &wg)
		}
	}

	wg.Wait()

	return int(cubics.Load())
}

func Task2(entries []Entry) int {
	perimetr, perimIndex, topLeft, downRight := buildPerimeter(entries)
	fmt.Println(topLeft, downRight)

	cubics := atomic.Uint64{}
	wg := sync.WaitGroup{}

	for y := topLeft.y; y < downRight.y+1; y++ {
		for x := topLeft.x; x < downRight.x+1; x++ {
			wg.Add(1)
			go func(x, y float64, wg *sync.WaitGroup) {
				defer wg.Done()
				if _, ok := perimIndex[Point2D{x, y}]; ok {
					cubics.Add(1)
					return
				}
				if PointInPolygonFast(Point2D{x, y}, perimetr) {
					cubics.Add(1)
				}
			}(x, y, &wg)
		}
	}

	wg.Wait()

	return int(cubics.Load())
}

func decodeHexIntoDec(hex string) int {
	dec, err := strconv.ParseInt(hex, 16, 64)
	if err != nil {
		panic(err)
	}
	return int(dec)
}

func buildPerimeter(entries []Entry) ([]Point2D, map[Point2D]struct{}, Point2D, Point2D) {
	perimeter := []Point2D{{0, 0}}
	perimIndex := map[Point2D]struct{}{{0, 0}: {}}

	topLeft, downRight := Point2D{math.Inf(1), math.Inf(1)}, Point2D{math.Inf(-1), math.Inf(-1)}

	for p := range entries {
		for i := 0; i < entries[p].Dist; i++ {
			lastPoint := perimeter[len(perimeter)-1]
			nextPoint := applyDirection(lastPoint, entries[p].Dir, entries[p].Dist)
			perimIndex[nextPoint] = struct{}{}
			if nextPoint.x > downRight.x {
				downRight.x = nextPoint.x
			}
			if nextPoint.y > downRight.y {
				downRight.y = nextPoint.y
			}
			if nextPoint.x < topLeft.x {
				topLeft.x = nextPoint.x
			}
			if nextPoint.y < topLeft.y {
				topLeft.y = nextPoint.y
			}
			perimeter = append(perimeter, nextPoint)
		}
	}

	return perimeter, perimIndex, topLeft, downRight
}

type Point2D struct {
	x, y float64
}

func PointInPolygonFast(target Point2D, polygon []Point2D) bool {
	if len(polygon) < 3 {
		return false
	}

	qPatt := [][]int{{0, 1}, {3, 2}}
	end := len(polygon) - 1
	prevPoint := Point2D{float64(polygon[end].x), float64(polygon[end].y)}
	prevPoint.x -= float64(target.x)
	prevPoint.y -= float64(target.y)
	prevQ := qPatt[boolToInt(prevPoint.y < 0)][boolToInt(prevPoint.x < 0)]

	w := 0

	for i := 0; i <= end; i++ {
		curPoint := polygon[i]

		curPoint.x -= float64(target.x)
		curPoint.y -= float64(target.y)

		q := qPatt[boolToInt(curPoint.y < 0)][boolToInt(curPoint.x < 0)]

		switch q - prevQ {
		case -3:
			w++
		case 3:
			w--
		case -2:
			if prevPoint.x*curPoint.y >= prevPoint.y*curPoint.x {
				w++
			}
		case 2:
			if !(prevPoint.x*curPoint.y >= prevPoint.y*curPoint.x) {
				w--
			}
		}

		prevPoint = curPoint
		prevQ = q
	}

	return w != 0
}

func boolToInt(b bool) int {
	if b {
		return 1
	}

	return 0
}

func numToDir(n int) string {
	if n == 0 {
		return "R"
	}
	if n == 1 {
		return "D"
	}
	if n == 2 {
		return "L"
	}
	if n == 3 {
		return "U"
	}

	panic("Unknown direction number" + strconv.Itoa(n))
}

func applyDirection(p Point2D, d string, step int) Point2D {
	if d == "R" {
		return Point2D{p.x + 1, p.y}
	}
	if d == "L" {
		return Point2D{p.x - 1, p.y}
	}
	if d == "D" {
		return Point2D{p.x, p.y + 1}
	}
	if d == "U" {
		return Point2D{p.x, p.y - 1}
	}

	panic("Unknown direction")
}

File: day18/main.py
from dataclasses import dataclass


@dataclass
class Point:
    x: int
    y: int

    def __hash__(self):
        return hash((self.x, self.y))


def bool_to_int(b):
    return int(b)


def point_in_polygon_fast(target, polygon):
    if len(polygon) < 3:
        return False

    q_patt = [[0, 1], [3, 2]]
    end = len(polygon) - 1
    prev_point = [polygon[end].x - target.x, polygon[end].y - target.y]
    prev_q = q_patt[bool_to_int(prev_point[1] < 0)
                    ][bool_to_int(prev_point[0] < 0)]

    w = 0

    for i in range(end + 1):
        cur_point = [polygon[i].x - target.x, polygon[i].y - target.y]
        q = q_patt[bool_to_int(cur_point[1] < 0)
                   ][bool_to_int(cur_point[0] < 0)]

        if q - prev_q == -3:
            w += 1
        elif q - prev_q == 3:
            w -= 1
        elif q - prev_q == -2:
            if prev_point[0] * cur_point[1] >= prev_point[1] * cur_point[0]:
                w += 1
        elif q - prev_q == 2:
            if not prev_point[0] * cur_point[1] >= prev_point[1] * cur_point[0]:
                w -= 1

        prev_point = cur_point
        prev_q = q

    return w != 0


def applyDirection(p: Point, d: str, step: int) -> Point:
    if d == 'R':
        return Point(p.x+1, p.y)
    if d == 'L':
        return Point(p.x-1, p.y)
    if d == 'D':
        return Point(p.x, p.y+1)
    if d == 'U':
        return Point(p.x, p.y-1)

# 76387


def task1():
    with open("day18/input") as f:
        parts = [(x[0], int(x[1]), x[2].strip("()"))
                 for x in [l.rstrip().split(" ") for l in f.readlines()]]
        perimetr = [Point(0, 0)]
        perimIndex = {Point(0, 0)}

        min_x = float('inf')
        min_y = float('inf')
        max_x = float('-inf')
        max_y = float('-inf')
        for p in parts:
            for i in range(p[1]):
                last_point = perimetr[-1]
                nextPoint = applyDirection(last_point, p[0], p[1])
                perimIndex.add(nextPoint)
                if nextPoint.x > max_x:
                    max_x = nextPoint.x
                if nextPoint.y > max_y:
                    max_y = nextPoint.y
                if nextPoint.x < min_x:
                    min_x = nextPoint.x
                if nextPoint.y < min_y:
                    min_y = nextPoint.y
                perimetr.append(nextPoint)
        print(min_x, max_x, min_y, max_y, len(perimIndex), len(perimetr))
        print(perimetr)

        cubics = 0
        total_cells = (max_x-min_x+1)*(max_y-min_y+1)
        i = 0
        for y in range(min_y, max_y+1):
            for x in range(min_x, max_x+1):
                if Point(x, y) in perimIndex:
                    cubics += 1
                    continue
                if point_in_polygon_fast(Point(x, y), perimetr):
                    cubics += 1
                i += 1
                if i % 1000 == 0:
                    print(i, total_cells, i*100/total_cells)

        print(cubics)

    return cubics


out1 = task1()
print(out1 == 76387, out1)
